### ISSUE TITLE
Re-usable test framework

### ISSUE THREAD

**Issue by JohnStarich** [spf13/afero#253]
While fixing some (small) inconsistencies in MemMapFs, I noticed there currently aren't generic tests for any given `afero.Fs`. I use several file systems from afero and a couple custom ones, so a package to test all of them for consistency could be invaluable to afero and the community.

Should one exist? If so, what would it look like? My hope is this issue can serve to explore how a "test framework" might look, and if it might have a home here in afero.

### Kinds of file systems

Looking at what's in afero today, there's a few shared use cases that could benefit from using the same tests.
Here's one attempt to categorize the built-in `afero.Fs` types and how we could test them:

| afero.Fs type | Test considerations                                                     |
|:--------------|:------------------------------------------------------------------------|
| OsFs          | basis for expected behavior                                             |
| MemMapFs      | same as OsFs                                                            |
| CacheOnReadFs | same as OsFs                                                            |
| CopyOnWriteFs | same as OsFs                                                            |
| BasePathFs    | same as OsFs, within a given directory                                  |
| RegexpFs      | same as OsFs, with wildcard `.*` names allowed                          |
| ReadOnlyFs    | same as OsFs, limited to read/stat tests                                |
| ZipFs         | same as OsFs, limited to read/stat tests                                |
| SftpFs        | (experimental) minimal read/write/stat tests                            |
| HttpFs        | need to verify if `HttpFs{}.Open` fits the signature of `afero.Fs.Open` |

Assuming the above table is a correct assessment, I see strong alignment with OsFs as the expected behavior.
As a first pass, I think it'd be practical to create an experimental `fstest` package and export one function:
```go
// Run verifies 'fs' behaves identically to *afero.OsFs.
func Run(t *testing.T, fs afero.Fs)
```

This function could cover 5 of the built-in file system types for where they overlap. For example:
```go
func TestBaseline(t *testing.T) {
	fstest.Run(t, NewMemMapFs())
	fstest.Run(t, NewCacheOnReadFs(NewMemMapFs(), NewMemMapFs(), 0))
	fstest.Run(t, NewCopyOnWriteFs(NewMemMapFs(), NewMemMapFs()))
	fstest.Run(t, NewBasePathFs(NewMemMapFs(), "/"))
	fstest.Run(t, NewRegexpFs(NewMemMapFs(), regexp.MustCompile(`.*`)))
}
```

We can even take that a step further, and add a read-only tester to `fstest`. This could cover at least 2 more built-in types:
```go
// RunReadOnly verifies 'fs' behaves identically to read operations in *afero.OsFs
func RunReadOnly(t *testing.T, fs afero.Fs)
```

A `fstest` package could be used internally for built-in types and by others to test custom `afero.Fs` implementations, too.

I imagine this could help catch bugs more easily. What are your thoughts on including such a framework in afero?

**Comment by JohnStarich** [comment]
For what it is worth, I've already started a reusable test suite like this for my own project. I'd be happy to drive the effort in afero if you think it's worth doing 🙂

**Comment by psadac** [comment]
I have done something similar here :  https://github.com/avfs/avfs ([here](https://github.com/avfs/avfs/test) more precisely) 
It may not be exactly what you are looking for, but you will find a bunch of tests for Linux mainly.
I don't have all the file systems present in Afero, however they are all tested with the same suite.

**Comment by JohnStarich** [comment]
Very cool, @psadac! It looks like it has really strong tests too. I'll give it a look 😄

**Comment by JohnStarich** [comment]
Initially, I decided to fork and maintain [that fork](https://github.com/JohnStarich/afero) with my `os` package compliance tests in another project.

However, I needed a stronger and more robust compliance test framework. I also thought it'd be nice to take advantage of Go 1.16's new interface-oriented FS design. I created [`hackpadfs`](https://github.com/hack-pad/hackpadfs) to meet those needs, with its own built-in compliance test framework [`fstest`](https://github.com/hack-pad/hackpadfs/tree/v0.1.0/fstest).

I think others may still benefit from a similar harness or integration in afero, but that's totally up you. 🙂 Since I've got a solution now, I'll go ahead and close.
